subsys/lorawan/lorawan.c

```diff
@@ -0,0 +1,601 @@
+/*
+ * Copyright (c) 2020 Manivannan Sadhasivam <mani@kernel.org>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+#include <errno.h>
+#include <net/lorawan.h>
+#include <zephyr.h>
+
+#include <LoRaMac.h>
+
+#ifdef CONFIG_LORAMAC_REGION_AS923
+	#define LORAWAN_REGION LORAMAC_REGION_AS923
+#elif CONFIG_LORAMAC_REGION_AU915
+	#define LORAWAN_REGION LORAMAC_REGION_AU915
+#elif CONFIG_LORAMAC_REGION_CN470
+	#define LORAWAN_REGION LORAMAC_REGION_CN470
+#elif CONFIG_LORAMAC_REGION_CN779
+	#define LORAWAN_REGION LORAMAC_REGION_CN779
+#elif CONFIG_LORAMAC_REGION_EU433
+	#define LORAWAN_REGION LORAMAC_REGION_EU433
+#elif CONFIG_LORAMAC_REGION_EU868
+	#define LORAWAN_REGION LORAMAC_REGION_EU868
+#elif CONFIG_LORAMAC_REGION_KR920
+	#define LORAWAN_REGION LORAMAC_REGION_KR920
+#elif CONFIG_LORAMAC_REGION_IN865
+	#define LORAWAN_REGION LORAMAC_REGION_IN865
+#elif CONFIG_LORAMAC_REGION_US915
+	#define LORAWAN_REGION LORAMAC_REGION_US915
+#elif CONFIG_LORAMAC_REGION_RU864
+	#define LORAWAN_REGION LORAMAC_REGION_RU864
+#elif
+	#error "Atleast one LoRaWAN region should be selected"
+#endif
+
+/* Use version 1.0.3.0 for ABP */
+#define LORAWAN_ABP_VERSION 0x01000300
+
+#define LOG_LEVEL CONFIG_LORAWAN_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(lorawan);
+
+K_SEM_DEFINE(mlme_confirm_sem, 0, 1);
+K_SEM_DEFINE(mcps_confirm_sem, 0, 1);
+
+K_MUTEX_DEFINE(lorawan_join_mutex);
+K_MUTEX_DEFINE(lorawan_send_mutex);
+
+static enum lorawan_datarate lorawan_datarate = LORAWAN_DR_0;
+static uint8_t lorawan_conf_msg_tries = 1;
+static bool lorawan_adr_enable;
+
+const char *status2str(int status)
```

Comment:
Is this function going to be used outside of lorawan.c?

Reply:
As long as 
https://github.com/zephyrproject-rtos/zephyr/pull/23664#discussion_r450027541